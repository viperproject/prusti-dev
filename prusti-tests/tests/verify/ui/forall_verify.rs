// compile-flags: -Pprint_desugared_specs=true -Pprint_typeckd_specs=true -Phide_uuids=true
// normalize-stdout-test: "[a-z0-9]{32}" -> "$(NUM_UUID)"
// normalize-stdout-test: "[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}" -> "$(UUID)"
// normalize-stdout-test: "\[[a-z0-9]{4}\]::" -> "[$(CRATE_ID)]::"

use prusti_contracts::*;

#[pure]
fn identity(x: i32) -> i32 { x }

#[ensures(forall(|x: i32| true))]
fn test1() {}

#[ensures(forall(|x: i32| identity(x) == x))]
fn test2() {}

#[ensures(forall(|x: i32| identity(x) == x + 1))]
fn test3() {}

#[ensures(exists(|x: i32| true))]
fn test4() {}

#[ensures(identity(1) == 1)]     // A witness.
#[ensures(exists(|x: i32| identity(x) == x))]
fn test5() {}

// TODO: Figure out why the error position is worse than for test3. I
// have checked the emitted Viper code (including the positions) and
// could not see any relevant differences.
#[ensures(exists(|x: i32| identity(x) == x + 1))]
fn test6() {}

fn main() {}
