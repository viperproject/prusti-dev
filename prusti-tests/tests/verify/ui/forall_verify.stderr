error: [Prusti: verification error] postcondition might not hold.
  --> $DIR/forall_verify.rs:17:27
   |
17 | #[ensures(forall(|x: i32| identity(x) == x + 1))]
   |                           ^^^^^^^^^^^^^^^^^^^^
   |
note: the error originates here
  --> $DIR/forall_verify.rs:18:1
   |
18 | fn test3() {}
   | ^^^^^^^^^^^^^

error: [Prusti: verification error] postcondition might not hold.
  --> $DIR/forall_verify.rs:31:1
   |
31 | fn test6() {}
   | ^^^^^^^^^^^^^

error: aborting due to 2 previous errors

